### PR TITLE
chat: add CHA-RL9 to explain "attaching piggyback"

### LINKED ISSUE #228: Presence spec points assume `ATTACHING` status is always caused by an `ATTACH` operation
> ** @(CHA-PR3d)@ @[Testable]@ If the room status is @Attaching@, then the @enter@ call will wait for the attach operation to succeed and then call the underlying presence @enter@ operation. It will throw an error if the attach fails.

(and similar points for other places we use channel presence operations)

This assumes that a room can only be in ATTACHING because of an ATTACH operation. This is not true — it can also be in this status because:

- a contributor entered ATTACHING and the transient disconnect timeout subsequently elapsed (CHA-RL4b7), or
- the RETRY operation put it into ATTACHED (CHA-RL5f)

How should the CHA-PR3d wait behave in this situation?

### PR DESCRIPTION
**Pull request by AndyTWF**
Closes #228

### CODE REVIEW

File: textile/chat-features.textile [hunk 1]
```diff
@@ -144,6 +144,10 @@ h4(#rooms-lifecycle-operations). Room Lifecycle Operations
 *** @(CHA-RL7a1)@ The @RETRY@ operation - an internal process.
 *** @(CHA-RL7a2)@ The @RELEASE@ operation.
 *** @(CHA-RL7a3)@ The @ATTACH@ and @DETACH@ operations have equal precedence.
+* @(CHA-RL9)@ Many operations in the Chat SDK will still attempt to proceed if the room is in an @ATTACHING@ status. However, their ability to proceed will depend on the subsequent status that the room takes. This specification point is defined here to avoid unnecessary repetition. It is testable via the points that refer to it.
```
Comment (lawrence-forooghian) [inline textile/chat-features.textile:147]: Could we mark these three spec points as Testable? Motivation being that I want to write some tests for this shared behaviour, and want to be able to able to make use of [the Swift repo’s annotations for attributing tests to a spec point](https://github.com/ably-labs/ably-chat-swift/blob/main/CONTRIBUTING.md#attributing-tests-to-a-spec-point). But the Swift repo’s script for turning these annotations into a spec coverage report emits a warning if you have an annotation for a spec point that isn’t marked as Testable (the idea being to catch places where we've implemented spec points that were subsequently deleted).

tl;dr: it would help the Swift repo’s tooling if a-c were marked as Testable here, but if there's a strong reason for them not to be then I can work around it

**Review by lawrence-forooghian — APPROVED** [review]

File: textile/chat-features.textile [hunk 2]
```diff
@@ -144,6 +144,10 @@ h4(#rooms-lifecycle-operations). Room Lifecycle Operations
...
+** @(CHA-RL9a)@ When the room status is @ATTACHING@ at the point of operation commencement, the caller will subscribe a one-time listener to the room status.
+** @(CHA-RL9b)@ If the next room status received by the caller is @ATTACHED@, the operation shall proceed as normal.
+** @(CHA-RL9c)@ If the next room status received is any other value, the operation must throw an error. The @ErrorInfo@ shall have the @RoomInInvalidState@ error code from the "chat-specific error codes":#error-codes, with a status code of @500@. The @cause@ field must contain any error associated with the room status change, if present.
```
Comment (sacOO7) [inline textile/chat-features.textile:150]: If it's any other value, like `DETACHED`, `SUSPENDED` or `FAILED` we are throwing error with code `RoomInInvalidState` = 102107.
Why don't we throw the same in the case of CHA-PR3f, CHA-PR3 or any other pre-op check that doesn't involve a status change

File: textile/chat-features.textile [hunk 3]
```diff
@@ -284,7 +288,7 @@ Presence allows chat room users to indicate to others that they're online, as we
 ** @(CHA-PR3c)@ This specification point has been removed.
 *** @(CHA-PR3c1) This specification point has been removed.
 *** @(CHA-PR3c2)@ This specification point has been removed.
-** @(CHA-PR3d)@ @[Testable]@ If the room status is @Attaching@, then the @enter@ call will wait for the attach operation to succeed and then call the underlying presence @enter@ operation. It will throw an error if the attach fails.
+** @(CHA-PR3d)@ @[Testable]@ If the room status is @Attaching@, then the operation shall enter a @CHA-RL9@ pending room status operation. If this succeeds (i.e the room becomes @ATTACHED@), the library must invoke the underlying @presence.enter()@ call.
 ** @(CHA-PR3e)@ @[Testable]@ If the room status is @Attached@, then the @enter@ call will invoke the underlying @presence.enter()@ call.
 ** @(CHA-PR3f)@ @[Testable]@ If the room status is @Detached@, an @ErrorInfo@ with code 40000, explaining that attach must be called first, will be thrown.
```
Comment (sacOO7) [inline textile/chat-features.textile:293]: We can throw the same here `RoomInInvalidState`

File: textile/chat-features.textile [hunk 4]
```diff
@@ -306,8 +310,8 @@ Presence allows chat room users to indicate to others that they're online, as we
 ** @(CHA-PR6b)@ This specification point has been removed.
 *** @(CHA-PR6b1)@ This specification point has been removed.
 *** @(CHA-PR6b2)@ This specification point has been removed.
-** @(CHA-PR6c)@ @[Testable]@ If the room status is @Attaching@, then the @get@ call will wait for the attach operation to succeed and then call the underlying presence @get@ operation. It will throw an error if the attach fails.
-** @(CHA-PR6d)@ @[Testable]@ If the room status is @Attached@, then the @get@ call will invoke the underlying @presence.enter()@ call.
+** @(CHA-PR6c)@ @[Testable]@ If the room status is @Attaching@, then the operation shall enter a @CHA-RL9@ pending room status operation. If this succeeds (i.e the room becomes @ATTACHED@), the library must invoke the underlying @presence.get()@ call.
+** @(CHA-PR6d)@ @[Testable]@ If the room status is @Attached@, then the @get@ call will invoke the underlying @presence.get()@ call.
 ** @(CHA-PR6e)@ @[Testable]@ If the room status is @Detached@, an @ErrorInfo@ with code 40000, explaining that attach must be called first, will be thrown.
 ** @(CHA-PR6f)@ @[Testable]@ For any other room status, an @ErrorInfo@ with code 40000 will be thrown.
```
Comment (sacOO7) [inline textile/chat-features.textile:316]: Or here `RoomInInvalidState`

File: textile/chat-features.textile (hunk 1 above)
Comment (AndyTWF) [reply textile/chat-features.textile:147]: Yep, can certainly do that!

File: textile/chat-features.textile (hunk 2 above)
Comment (AndyTWF) [reply textile/chat-features.textile:150]: In principle I think this is reasonable - tbf when we wrote the spec point, `RoomInInvalidState` wasn't an error yet.

Though the thing we need to be careful of is that `RoomInInvalidState` is supposed to be a 5xx status code, whereas what we really want for CHA-PR3f is more like a 4xx, because its the client doing something wrong, I'll have a think about it.

Comment (AndyTWF) [reply textile/chat-features.textile]: Change made

File: textile/chat-features.textile [hunk 5]
```diff
@@ -368,10 +375,11 @@ Typing Indicators allow chat room users to indicate to others that they are typi
 ** @(CHA-T2b)@ This specification point has been removed.
 *** @(CHA-T2b1)@ This specification point has been removed.
 *** @(CHA-T2b2)@ This specification point has been removed.
-** @(CHA-T2c)@ @[Testable]@ If the room status is @ATTACHING@, then the @get@ call shall wait for the attach operation to succeed and then call the underlying presence @get@ operation. It must throw an error if the attach fails.
+** @(CHA-T2c)@ @[Testable]@ If the room status is @Attaching@, then the operation shall enter a @CHA-RL9@ pending room status operation. If this succeeds (i.e the room becomes @ATTACHED@), the library must invoke the underlying @presence.get()@ call.
```
Comment (sacOO7) [inline textile/chat-features.textile:378]:
```suggestion
** @(CHA-T2c)@ @[Testable]@ If the room status is @Attaching@, then the operation should wait as per @CHA-RL9a@, once RoomStatus is @Attached@ as per @CHA-RL9b@ then the library must invoke the underlying @presence.get()@ call, otherwise throw error as per @CHA-RL9c@
```

File: textile/chat-features.textile [hunk 6]
```diff
@@ -344,10 +350,11 @@ Presence allows chat room users to indicate to others that they're online, as we
 ** @(CHA-PR6b)@ This specification point has been removed.
 *** @(CHA-PR6b1)@ This specification point has been removed.
 *** @(CHA-PR6b2)@ This specification point has been removed.
-** @(CHA-PR6c)@ @[Testable]@ If the room status is @Attaching@, then the @get@ call will wait for the attach operation to succeed and then call the underlying presence @get@ operation. It will throw an error if the attach fails.
-** @(CHA-PR6d)@ @[Testable]@ If the room status is @Attached@, then the @get@ call will invoke the underlying @presence.enter()@ call.
-** @(CHA-PR6e)@ @[Testable]@ If the room status is @Detached@, an @ErrorInfo@ with code 40000, explaining that attach must be called first, will be thrown.
-** @(CHA-PR6f)@ @[Testable]@ For any other room status, an @ErrorInfo@ with code 40000 will be thrown.
+** @(CHA-PR6c)@ @[Testable]@ If the room status is @Attaching@, then the operation shall enter a @CHA-RL9@ pending room status operation. If this succeeds (i.e the room becomes @ATTACHED@), the library must invoke the underlying @presence.get()@ call.
```
Comment (sacOO7) [inline textile/chat-features.textile:353]:
```suggestion
** @(CHA-PR6c)@ @[Testable]@ If the room status is @Attaching@, then the operation should wait as per @CHA-RL9a@, once RoomStatus is @Attached@ as per @CHA-RL9b@ then the library must invoke the underlying @presence.get()@ call, otherwise throw error as per @CHA-RL9c@
```

File: textile/chat-features.textile [hunk 7]
```diff
@@ -322,20 +326,22 @@ Presence allows chat room users to indicate to others that they're online, as we
...
 * @(CHA-PR10)@ Users may update their presence data. The behaviour depends on the current room status, as presence operations in a Realtime Client cause implicit attaches.
 ** @(CHA-PR10a)@ @[Testable]@ Users may choose to update their presence data, optionally providing custom data to update with. The overall presence data must retain the format specified in @CHA-PR2@.
 ** @(CHA-PR10b)@ This clause has been replaced by @CHA-PR10c - CHA-PR10g@.
 ** @(CHA-PR10c)@ This specification point has been removed.
 *** @(CHA-PR10c1)@ This specification point has been removed.
 *** @(CHA-PR10c2)@ This specification point has been removed.
-** @(CHA-PR10d)@ @[Testable]@ If the room status is @Attaching@, then the @update@ call will wait for the attach operation to succeed and then call the underlying presence @update@ operation. It will throw an error if the attach fails.
+** @(CHA-PR10d)@ @[Testable]@ If the room status is @Attaching@, then the operation shall enter a @CHA-RL9@ pending room status operation. If this succeeds (i.e the room becomes @ATTACHED@), the library must invoke the underlying @presence.update()@ call.
```
Comment (sacOO7) [inline textile/chat-features.textile:340]:
```suggestion
** @(CHA-PR10d)@ @[Testable]@ If the room status is @Attaching@, then the operation should wait as per @CHA-RL9a@, once RoomStatus is @Attached@ as per @CHA-RL9b@ then the library must invoke the underlying @presence.update()@ call, otherwise throw error as per @CHA-RL9c@
```

**Comment by sacOO7** [comment]
This is also applicable on sending ephermal reactions via send method right?

**Comment by AndyTWF** [comment]
> This is also applicable on sending ephermal reactions via send method right?

Yes, but I decided not to alter the spec here because there's currently a DR open that will make that spec point entirely redundant.

File: textile/chat-features.textile [hunk 8]
```diff
@@ -144,7 +144,11 @@ h4(#rooms-lifecycle-operations). Room Lifecycle Operations
...
+* @(CHA-RL9)@ Many operations in the Chat SDK will still attempt to proceed if the room is in an @ATTACHING@ status. However, their ability to proceed will depend on the subsequent status that the room takes. This specification point is defined here to avoid unnecessary repetition. It is primarily testable via the points that refer to it.
+** @(CHA-RL9a)@ @[Testable]@ When the room status is @ATTACHING@ at the point of operation commencement, the caller will subscribe a one-time listener to the room status.
+** @(CHA-RL9b)@ @[Testable]@ If the next room status received by the caller is @ATTACHED@, the operation shall proceed as normal.
+** @(CHA-RL9c)@ @[Testable]@ If the next room status received is any other value, the operation must throw an error. The @ErrorInfo@ shall have the @RoomInInvalidState@ error code from the "chat-specific error codes":#error-codes, with a status code of @500@. The @cause@ field must contain any error associated with the room status change, if present.
+>>>>>>> db59a5bd (chat: add CHA-RL9 to explain "attaching piggyback")
```
[the remark this answers is not in the document]
Comment (AndyTWF) [reply textile/chat-features.textile:151]: Didn't show up in the IDE, it's gone now though.

File: textile/chat-features.textile (hunk 7 above)
Comment (AndyTWF) [reply textile/chat-features.textile]: Will mark this as resolved and continue discussion on the one I've left open

File: textile/chat-features.textile (hunk 6 above)
Comment (AndyTWF) [reply textile/chat-features.textile]: Will mark this as resolved and continue discussion on the one I've left open

File: textile/chat-features.textile (hunk 5 above)
Comment (AndyTWF) [reply textile/chat-features.textile]: Will mark this as resolved and continue discussion on the one I've left open

File: textile/chat-features.textile [hunk 9]
```diff
@@ -322,20 +326,22 @@ Presence allows chat room users to indicate to others that they're online, as we
 ** @(CHA-PR3c)@ This specification point has been removed.
 *** @(CHA-PR3c1) This specification point has been removed.
 *** @(CHA-PR3c2)@ This specification point has been removed.
-** @(CHA-PR3d)@ @[Testable]@ If the room status is @Attaching@, then the @enter@ call will wait for the attach operation to succeed and then call the underlying presence @enter@ operation. It will throw an error if the attach fails.
+** @(CHA-PR3d)@ @[Testable]@ If the room status is @Attaching@, then the operation shall enter a @CHA-RL9@ pending room status operation. If this succeeds (i.e the room becomes @ATTACHED@), the library must invoke the underlying @presence.enter()@ call.
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply textile/chat-features.textile]: Sure, no issues! 
The only confusion I see is with the statement `then the operation shall enter a @CHA-RL9@ pending room status operation`. 
Can we make it explicit, that it should `wait for the ATTACH operation` and not `enter it` wdyt

[the remark this answers is not in the document]
Comment (AndyTWF) [reply textile/chat-features.textile]: Reworded :)

[the remark this answers is not in the document]
Comment (sacOO7) [reply textile/chat-features.textile]: Thanks, LGTM

**Review by sacOO7 — APPROVED** [review]
LGTM

**Comment by lawrence-forooghian** [comment]
I want to get this one merged so that I can implement it in Swift; have rebased and will merge.

**Comment by sacOO7** [comment]
@AndyTWF Btw, do we not check for ATTACHED state for presence leave as per CHA-PR4 ?

**Comment by AndyTWF** [comment]
> @AndyTWF Btw, do we not check for ATTACHED state for presence leave as per CHA-PR4 ?

Yes - will add.

**Comment by sacOO7** [comment]
The same goes for CHA-T4a1 ( typing presence enter ) and CHA-T4a2 ( typing presence leave )
